### PR TITLE
feat(layer): allow sublayer and variable reassignment after build

### LINKED ISSUE #18601: Consider allowing sublayer reassignment
We are having a lot of discussion around lora/quantization and other PEFT strategies that require replacing layers (often dense layers), with parameter efficient replacements.

In torch, a `nn.Module` will track submodules by name. So you can run `module.sub_module = new_sub_module` without issue. The old child will be booted out for the new. Same for `tf.Module`.

This is not the case with Keras layers. Tracking is currently "append only," not by attr name, and locked after `build()`. To work around this for a LoRA implementation, we currently do the following.

```python
def replace(parent, name, replacement):
    locked = parent._tracker.locked
    parent._tracker.locked = False
    target = getattr(parent, name)
    setattr(parent, name, replacement)
    parent._layers[parent._layers.index(target)] = replacement
    parent._tracker.locked = locked
```

We should consider whether we want to allow this for Keras layers with public APIs. Potentially by extending the tracker to track by attr name.

### PR DESCRIPTION
**Pull request by rstar327**
## Description

Allows reassigning an existing tracked sublayer or `Variable` on a Keras `Layer` after it has been built. Fixes #18601.

### Motivation

Parameter-efficient fine-tuning recipes (LoRA, quantization, adapter insertion) need to swap a built layer's child for a parameter-efficient variant. Today this requires reaching into private tracker internals:

```python
def replace(parent, name, replacement):
    locked = parent._tracker.locked
    parent._tracker.locked = False
    target = getattr(parent, name)
    setattr(parent, name, replacement)
    parent._layers[parent._layers.index(target)] = replacement
    parent._tracker.locked = locked
```

After this change the same operation is just:

```python
parent.sub_layer = replacement
```

### Behavior

`Layer.__setattr__` now detects the case where the name already holds a tracked object of the same kind (Layer, Variable, Metric, SeedGenerator) and the tracker is locked. Instead of failing on `add_to_store`, it calls the tracker's existing `replace_tracked_value`, which swaps the entry in place and updates the id index.

Adding a *brand new* sublayer or variable after `build()` is still rejected — only reassignment of an already-tracked slot is allowed. The original "you cannot add new elements of state to a layer that is already built" error path is untouched.

### Tests

Added `test_sublayer_can_be_reassigned_after_build` and `test_variable_can_be_reassigned_after_build` in `keras/src/layers/layer_test.py`. Both verify:
- The reassignment doesn't raise
- The new object is returned by attribute access
- The new object appears exactly once in the tracker's store (no stale entries, no duplicates)
- The new object is the one used by the forward pass

Full `layer_test.py` suite (61 tests, 1 skip) still passes on tensorflow, jax, torch, numpy backends locally.

## Contributor Agreement

**Please review our [AI-Assisted Contribution Policy](../CONTRIBUTING.md#ai-assisted-contribution-policy) and check all boxes below before submitting your PR for review:**

- [x] I am a human, and not a bot.
- [x] I will be responsible for responding to review comments in a timely manner.
- [x] I will work with the maintainers to push this PR forward until submission.

**Note:** Failing to adhere to this agreement may result in your future PRs no longer being reviewed.

### CODE REVIEW

**Review by hertschuh — CHANGES_REQUESTED** [review]
Unfortunately, this is not enough, for a reason that has to do with how the tracker works today.

The tracker doesn't keep the mapping between names and the mapped value.

So if you see a value in the store, you evict it. However, the same value could be tracked via multiple attributes, which is actually a common pattern:

```python
self.dense1 = Dense(10)
self.dense2 = Dense(5)
self.all_layers = [self.dense1, self.dense2]
```

So you may be evicting something that still needs to be tracked.

It might look like a corner case, but it will probably break existing users.

**Comment by rstar327** [comment]
Good catch, thanks — you're right that replacing in the store breaks the multi-attribute case. Reworked it: on reassignment I now track the new value, then untrack the old one **only if no other attribute still references it** (a scan over direct attributes + tracked containers, treating layers/variables as leaves). So in your example `old_dense1` stays tracked because `self.all_layers` still holds it, while a genuinely orphaned value is still cleaned up. Adding brand-new state after build is still rejected. Added a test for the exact `self.all_layers = [self.dense1, self.dense2]` pattern; `layer_test.py` passes on all backends.
